src/Symfony/Component/FeatureToggle/FeatureCollection.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\FeatureToggle;
+
+use Psr\Container\ContainerInterface;
+
+/** @implements \IteratorAggregate<int, Feature> */
+final class FeatureCollection implements ContainerInterface, \IteratorAggregate
+{
+    /** @var array<string, Feature>|null */
+    private array|null $features = null;
+
+    /** @var array<iterable<Feature>|(\Closure(): iterable<Feature>)> */
+    private array $featureProviders = [];
+
+    /**
+     * @param iterable<Feature> $features
+     */
+    public function __construct(iterable $features)
+    {
+        $this->append($features);
+    }
+
+    /**
+     * @param iterable<Feature>|(\Closure(): iterable<Feature>) $features
+     */
+    private function append(iterable|\Closure $features): void
+    {
+        $this->featureProviders[] = $features;
+        $this->features = null;
+    }
+
+    /**
+     * @param iterable<Feature>|(\Closure(): iterable<Feature>) $features
+     */
+    public function withFeatures(iterable|\Closure $features): self
+    {
+        $this->append($features);
+
+        return $this;
+    }
+
+    /**
+     * @phpstan-assert-if-true null $this->features
+     */
+    private function compile(): void
+    {
+        if (null !== $this->features) {
+            return;
+        }
+
+        $this->features = [];
+
+        foreach ($this->featureProviders as $featureProvider) {
+            if (is_callable($featureProvider)) {
+                $featureProvider = $featureProvider();
+            }
+
+            foreach ($featureProvider as $feature) {
+                $this->features[$feature->getName()] = $feature;
+            }
+        }
+    }
+
+    public function has(string $id): bool
+    {
+        $this->compile();
+        return array_key_exists($id, $this->features);
+    }
+
+    /**
+     * @throws FeatureNotFoundException If the feature is not registered in this provider.
+     */
+    public function get(string $id): Feature
+    {
+        $this->compile();
```

Comment:
This implementation compiling all features from various providers on usage makes it impossible to build any optimization where we would build the list in a compiler pass during the cache warmup.
It also forces to instantiate all features to be able to use a single one.

IMO, a better API would be to inject a PSR ContainerInterface in the constructor. This way, the wiring in the framework could inject a ServiceLocator, which would lazy-load the Feature object that we need and no other.

For the iterator, I see 2 options:
- don't make it iterable at all (if we can)
- inject a list of feature ids in that class so that it could iterate over them and lazy-load the Feature objects associated with them.

Reply:
> makes it impossible to build any optimization where we would build the list in a compiler pass during the cache warmup.

Actually  this is what is currently dumped in the cache :
```php
$instance->withFeatures((new \Symfony\Component\FeatureToggle\Provider\InMemoryProvider([
    new \Symfony\Component\FeatureToggle\Feature('my-feature', 'Desc feature 1', true, 
        new \Symfony\Component\FeatureToggle\Strategy\PriorityStrategy([
            new \Symfony\Component\FeatureToggle\Strategy\RequestQueryStrategy('my_query'), 
            new \Symfony\Bundle\FeatureToggleBundle\Strategy\RequestStackAttributeStrategy('some_attribute', ($container->services['request_stack'] ??= new \Symfony\Component\HttpFoundation\RequestStack())),
        ])
    )
]))->provide(...));
```

I agree all features (in case of `InMemoryProvider`) are always instanciated but I don't see what else could be optimized during cache warmup.

Also take into account that feature names can come from SAAS as well. So unless we accept doing http calls during cache warmup it won't be possible to build the full list of available features. Also doing this during cache warmup would prevent people from updating the list on those SAAS providers without having to refresh symfony's cache.

> IMO, a better API would be to inject a PSR ContainerInterface in the constructor. This way, the wiring in the framework could inject a ServiceLocator, which would lazy-load the Feature object that we need and no other.

Like I said, this implies to have access to all the features which is currently not possible unless doing http calls or else. Lazy loading the Feature Object why not but it is a very light object. What would need lazy is the strategy classes IMO

Reply:
If the feature comes from a SaaS tool, loading the full list might still not be the right API (and which strategy would they use then ?). I'm not convinced the existing API is good for those SaaS providers either.

Reply:
We got inspiration with how they do it in their own package and the HTTP API they exposes. They don't provide several (one to list, one to determine and so on), they pull everything locally. They continue to have some remote call to store per request which user was given which answer. What they do is on first call they put everything in cache. ATM this component does not provide full integration with such external tools. Yes there will be improvement to bring these capabilities "10% of users" for example. ATM this can already be plugged in with those tools by wrapping their actual plugins.

Regarding which strategies they will choose this has to be discussed because this can go several ways. We implement those logic in our own micro components (like notifier) or we simply bridge with their existing packages.

The strategies they offer always relies on a user list or percentage or globally available but is not as flexible as what this component tries to bring to the table (switch based on query, header, request attribute, etc etc).

For now what I can provide as a better implementation is to not compile everything right from the start but iterate over providers and stop as soon as one does provide the feature.

Reply:
I've done an update on the FeatureCollection class. What could be done to improve further would be to add a `has` and a `get` method on providers. Regarding the InMemory it is an easy one and could avoid to instanciate feature if not needed. For external ones it would up to each of them to optimize things. For gitlab for example : 
- fetch from https://docs.gitlab.com/ee/api/feature_flags.html#list-feature-flags-for-a-project 
- store it in cache or keep it in memory (don't know yet which will be better suited)
- and simply extract feature names from it to implement the has / get methods.

WDYT ?

Reply:
Providers were totally reworked. Let me know what you think.